### ISSUE TITLE
Fundamental memory limit

### ISSUE THREAD

**Issue by aelanman** [RadioAstronomySoftwareGroup/healvis#24]
Parallelization over the time axis was necessary to accelerate simulations of lots of LST, but it comes at a cost in memory.

Let:
- `Nprocs` = The number of parallel processes
- `Npix_sel` = The number of pixels in the selected field of view
- `Nfreqs` = The number of frequencies in the chosen passband

At each time step, the observatory calculates `lmn` coordinates from the vector pixel centers and the pointing center. This means that, in each subprocess, there are 8 separate arrays of size `Npix_sel` being generated simultaneously: `x, y, z, az, za, l, m, n`. Granted, some of these are only used once and so Python's garbage collection might get rid of them, but at least l,m,n are kept throughout the visibility calculation step.

So it's possible for the total memory footprint to reach `Nprocs * Npix_sel * Nfreqs * 8 * sizeof(float64)`.

This is only really a problem when the FoV is large, which is recommended when simulating foregrounds.

The partial frequency method already helps with this, but there are a few other things we might try:
- Put the pixel center vectors (x,y,z) in shared memory. This avoids recalculating them on each process as well as reduplication of those variables.
- Replace this vector lmn calculation with a generator function. Question -- Would this still work within an einsum?
- Option to parallelize over pixels instead of time if the number of pixels in the FoV is substantially larger than the number of times.
- Or, as a variation, nested parallelization: From the given `Nprocs`, specify `Nprocs_time` and `Nprocs_pixel` such that `Nprocs_time * Nprocs_pixel = Nprocs`. The `make_visibilities` function then spawns `Nprocs_time` processes which each spawn `Nprocs_pixel` processes which work on subsections of the FoV. The total memory footprint scales with `Nprocs_time` instead of `Nprocs`.

This is a deep issue that will take some effort to resolve. The main speed advantage of healvis over pyuvsim comes from the vectorized position and visibility calculations, but this is complicated by the parallelization.

**Comment by nkern** [comment]
Some interesting points here: I don't know if I have any intuition for some of the proposed solutions, but generally because we generally simulate on the order of hundreds if not order thousand frequency channels, the `run_simulation_partial_freq` in princple does allow us to cut down on instantaneous memory loads by order 100-1000. For any reasonable simulation setup that seems like that should be more than enough to put a run under most cluster resource limits. LMK if you find that even with partial frequency simulation that memory loads are an issue.

**Comment by aelanman** [comment]
Certainly! This isn't a show-stopper, but it is a flaw worth documenting for future development.

Ideally it would be possible to just start the simulation with some set of available resources and have it decide how best to balance them to run the task --- if necessary, trading speed for memory. At a base level, running with `Nprocs=1` minimizing the memory footprint while maximizing runtime. For example, if we know the memory footprint increases by a factor of 75% for each additional processor, then it should be possible to figure out before the simulation starts if the full bandwidth/FoV will exceed available memory. From there, it should be able to decide how to split the bandwidth for partial frequency simulation, etc.

The third proposed solution I gave is a version of this idea. To sketch the steps:

1. Check available memory `Mtot` and the size of the skymodel section `Msky = Npix * Nfreqs * sizeof(float)`
2. We will split `Nprocs` into `Nprocs = Nskysec * Ntsec`, where `Nskysec` and `Ntsec` are the number of sky "sections" and time sections (ie., splitting the pixel and time axes, respectively).
3. The anticipated memory footprint is `M = Ntsec * 8*Msky * FoV/4π`, where FoV is the solid angle of the field of view. It's Ntsec, because at any given time that's how many copies of the FoV are being held in memory.
4. If `M/Mtot < 1`, then there is sufficient memory: `Ntsec = Nprocs, Nskysec=1`
5. Otherwise, find a different factorization until `M/Mtot < 1`. If it can't be done, then there simply isn't enough memory for this job (or go further and try splitting the frequencies using partial_freq sim...)
6. The `make_visibilities` method spawns `Nproc` processes and assign sections of the time array and sky array to these based on how `Ntsec` and `Nskysec` are set.

Another option, come to think of it... it's only for clarity's sake that we calculate az/za and lmn up front like that. We could instead write the `vec -> lmn` operation into the visibility calculation itself, given the pointing center. We then keep the pixel center vectors in shared memory since they don't change with time. I'm not sure how much that might slow things down, but that would reduce the memory footprint.

Sorry, I'm just kindof thinking out loud at this point... This is a long-term goal of `pyuvsim` that I'm projecting here. The expected behavior is that with more processors performance should improve, but that is not entirely true because of the memory inflation with `Nprocs`.
